### PR TITLE
chore: Check OAuth token when workspace starts

### PR DESCRIPTION
**Pull request by vinokurig**
Signed-off-by: Igor Vinokur <ivinokur@redhat.com>

### What does this PR do?
* Set a flag to the browsers `SessionStorageService` when an OAuth authorisation is requested.
* When a factory is created and stoped and started again, check if the flag is set.
* If the flag is set, send a request to `che-server` to check the OAuth token
* If `che-server` returns an error, open an authentication page to authenticate.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/21640

### CODE REVIEW

File: packages/dashboard-backend/src/workspace.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2022 Red Hat, Inc.
```
Comment (akurinnoy) [inline packages/dashboard-backend/src/workspace.ts:2]: This is what we need to update, but right now, the linter expects:

```suggestion
 * Copyright (c) 2018-2021 Red Hat, Inc.
```

File: packages/dashboard-frontend/src/store/Workspaces/index.ts [hunk 2]
```diff
@@ -176,10 +183,38 @@ export const actionCreators: ActionCreators = {
   startWorkspace:
     (workspace: Workspace, params?: ResourceQueryParams): AppThunk<KnownAction, Promise<void>> =>
     async (dispatch, getState): Promise<void> => {
+      const openOauthPage = (oauthUrl: string, namespace: string, workspaceName: string) => {
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/store/Workspaces/index.ts:186]: Don't you mind extracting this function and the similar one from the factory loader ([openAuthPage](https://github.com/eclipse-che/che-dashboard/pull/640/files#diff-497e6c2dff7f716c2f404dcddd0c83667f010ded8ac6eb4474d445dc45eab5dfR207-R225)) into a separate module (service)?

File: packages/dashboard-frontend/src/store/Workspaces/index.ts [hunk 3]
```diff
@@ -176,10 +183,38 @@ export const actionCreators: ActionCreators = {
...
+          workspace.devfile.projects
+        ) {
+          SessionStorageService.update(SessionStorageKey.AUTHENTICATION_STATUS, 'started');
+          const project = workspace.devfile.projects.find(p => (p.name = workspace.projects[0]));
+          if (project && project.git) {
+            await WorkspaceClient.restApiClient.refreshFactoryOauthToken(
+              project.git.remotes.origin,
+            );
+          }
+        }
```
Comment (olexii4) [inline packages/dashboard-frontend/src/store/Workspaces/index.ts:217]: @vinokurig I would also like to extract these lines into an OAuth service, like https://github.com/eclipse-che/che-dashboard/pull/640/files#r985524100

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Fetch/Devfile/index.tsx [hunk 4]
```diff
@@ -186,6 +186,7 @@ class StepFetchDevfile extends AbstractLoaderStep<Props, State> {
     });
 
     try {
+      SessionStorageService.update(SessionStorageKey.AUTHENTICATION_STATUS, 'started');
```
Comment (olexii4) [inline packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Fetch/Devfile/index.tsx:189]: I think this line should be a method in the OAuth service.

File: packages/dashboard-frontend/src/store/Workspaces/index.ts (hunk 2 above)
Comment (vinokurig) [reply packages/dashboard-frontend/src/store/Workspaces/index.ts]: done

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Fetch/Devfile/index.tsx (hunk 4 above)
Comment (vinokurig) [reply packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Fetch/Devfile/index.tsx]: done

**Comment by vinokurig** [comment]
@akurinnoy @olexii4 Any updates?

**Comment by vinokurig** [comment]
To test it you have to 
* Configure GitHub OAuth https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-2-for-github/
* Create a workspace from a GitHub repo url, apply the GitHub authentication.
* Stop the workspace.
* Revoke the GitHub token in your GitHub account.
* Start the worksapce.
See: Github asks to authenticate again.

**Review by akurinnoy — APPROVED** [review]
Tested, and it works as expected!

**Pull request merged**